[flutter_tools] Revamp observatory port discovery logic on Fuchsia

michaellee8 opened the pull request:
- Fix dart observatory discovery protocol on Fuchsia.

Further work on https://github.com/flutter/flutter/issues/83609

This PR is now ready for review.

This should be merged after https://github.com/flutter/flutter/pull/85752.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

michaellee8 committed:
feat: implement session_control for starting app in workstation build

michaellee8 committed:
feat: teach flutter_tool to close ermine app by restarting it

chaselatta commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -657,11 +727,26 @@ class FuchsiaDevice extends Device {
+  ///
+  /// If the device is running a workstation build, `session_control` should be
+  /// used to launch apps, otherwise `tiles_ctl` should be used.
+  Future<bool> isWorkstation() async {
```
Please don't hardcode any product configuration details into the flutter tool because it should be able to work with any product configuration.

michaellee8 replied:
You are right. I am thinking of an alternative way to detect whether we should use tiles or session. Maybe detecting the existence of a session_control binary would work?

chaselatta commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -657,11 +727,26 @@ class FuchsiaDevice extends Device {
+    if (result.exitCode != 0) {
+      return false;
+    }
+    if (result.stdout.contains('workstation')) {
```
This is not part of the platform API and should not be relied upon. This will break if workstation changes their product name and it will also not scale if we add other products.

michaellee8 committed:
fix: wip

michaellee8 commented:
I have got some progress here, you can see https://youtu.be/XCaNaxVI1hI to see my working code.

michaellee8 committed:
fix: working flutter run on fuchsia workstation build

This is the version before clean up, for the record.

michaellee8 committed:
fix: prevent FlutterIsolateDiscoveryProtocol from using creating DDS instance

Before this change, FlutterIsolateDiscoveryProtocol identifies the
correct vm service port for the flutter app we are debugging by
connecting to the vm service, creating a Flutter DDS instance, and
then query whether the the vm is serving our required isolate. This
method has indeed caused some problems, one of them is that the FlutterIsolateDiscoveryProtocol passes the port to the DDS instance back to startApp, in which ResidentRunner tries to create another
DDS instance on top of the existing DDS instance. This causes the
flutter_tools to crash since the underlying vm service has already
been instructed to enter Single Client Mode by the first DDS, hence
the second DDS will not be able to connect.

The current implementation instead queries the flutter apps running
under each runner by querying the /hub. Although implementation does
use the currently hub v1 protocol which is currently under
deprecation, it does not involves creating extra VM services hence
less error prone. I have tried to use the /hub-v2 protocol instead
but have met several bugs that made it not possible to be used
to complete this task. A TODO has been added to migrate to the hub-v2
protocol when possible.

michaellee8 committed:
fix: prevent FlutterIsolateDiscoveryProtocol from using creating DDS instance

Before this change, FlutterIsolateDiscoveryProtocol identifies the
correct vm service port for the flutter app we are debugging by
connecting to the vm service, creating a Flutter DDS instance, and
then query whether the the vm is serving our required isolate. This
method has indeed caused some problems, one of them is that the
FlutterIsolateDiscoveryProtocol passes the port to the DDS instance back to
startApp, in which ResidentRunner tries to create another
DDS instance on top of the existing DDS instance. This causes the
flutter_tools to crash since the underlying vm service has already
been instructed to enter Single Client Mode by the first DDS, hence
the second DDS will not be able to connect.

The current implementation instead queries the flutter apps running
under each runner by querying the /hub. Although implementation does
use the currently hub v1 protocol which is currently under
deprecation, it does not involves creating extra VM services hence
less error prone. I have tried to use the /hub-v2 protocol instead
but have met several bugs that made it not possible to be used
to complete this task. A TODO has been added to migrate to the hub-v2
protocol when possible.

This version is verified to have flutter run working with the
qemu-x64 and workstation.qemu-x64-release build on emulator.

michaellee8 commented:
A noteworthy commit message here, it documents why did I changed the implementation of FuchsiaIsolateDiscoveryProtocol

fix: prevent FlutterIsolateDiscoveryProtocol from using creating DDS instance

Before this change, FlutterIsolateDiscoveryProtocol identifies the correct vm service port for the flutter app we are debugging by
connecting to the vm service, creating a Flutter DDS instance, and then query whether the the vm is serving our required isolate. This method has indeed caused some problems, one of them is that the FlutterIsolateDiscoveryProtocol passes the port to the DDS instance back to startApp, in which ResidentRunner tries to create another DDS instance on top of the existing DDS instance. This causes the flutter_tools to crash since the underlying vm service has already been instructed to enter Single Client Mode by the first DDS, hence the second DDS will not be able to connect.

The current implementation instead queries the flutter apps running under each runner by querying the /hub. Although implementation does use the currently hub v1 protocol which is currently under deprecation, it does not involves creating extra VM services hence less error prone. I have tried to use the /hub-v2 protocol instead but have met several bugs that made it not possible to be used to complete this task. A TODO has been added to migrate to the hub-v2 protocol when possible.

This version is verified to have flutter run working with the qemu-x64 and workstation.qemu-x64-release build on emulator.

michaellee8 committed:
refactor: cleaning up some dead code

michaellee8 committed:
wip: debugging tests

michaellee8 committed:
wip: add device tests

michaellee8 committed:
test: write tests for session framework detection and observatory discovery

michaellee8 committed:
refactor: fix formatting

michaellee8 commented:
@chaselatta The coding has been done with relavant tests written and unneccessary changes rolled back so I will mark it ready for review now. Please review it. This is a rather larger PR so you may want to find other reviewers. Thanks for help in advance.

michaellee8 commented:
I can see some tool_tests failing, but I am not sure if that was my problem. I don't have code I changed are related to those failing tests.

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -593,7 +630,8 @@ class FuchsiaDevice extends Device {
 
   /// [true] if the current host address is IPv6.
-  bool get ipv6 => _ipv6 ??= isIPv6Address(id);
+  //  Strip interface name to avoid issues
```
Can this comment be more specific? Also @chaselatta, in light of https://github.com/flutter/flutter/pull/85341, I see that there are some sensitivities around the addresses that are used in different contexts, so I wanted to call this line to your attention for review.

michaellee8 replied:
The problem is that dart's ipv6 parsing logic assumes that ipv6 address are hex numbers seperated by `:`. It is unable to accept an ipv6 address that is suffixed by an interface name, say `[::1%qemu]`. Therefore in order to avoid worksraound this inconsistency, I just strip that interface component of an ipv6 address away.

michaellee8 replied:
Documented in 77e22ad50b0508d0b1682f7b9e437bb3fd224ae3, I will leave this unresolved in case you want more explaination or perharps another issue on dart sdk for this.

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -621,8 +659,13 @@ class FuchsiaDevice extends Device {
 
-  /// List the ports currently running a dart observatory.
-  Future<List<int>> servicePorts() async {
+  /// List the ports currently running a dart observatory and the component url
```
If possible, I'd suggest pulling the changes to the port discovery logic out into a separate PR, with an issue filed first that describes the problem that it's solving.

OTOH, if it isn't possible to use `session_control` without it, then it would have to stay.

michaellee8 replied:
If I seperate the changes in port discovery logic to a seperate PR, it seems that any usage of workstation related code will cause flutter_tools to crash. I am not sure if just adding session_tools will pass the tests through. 

Actually without the changes in servicePorts() it is not just `workstation.qemu-x64-release` build will crash `flutter_tools`, even `qemu-x64` build will not be able to connect to dart observatory. I don't understand the root cause of what crashes the original port discovery logic in workstation build, but I think it is because the original port discovery logic does not work correctly when there are multiple instances of flutter_runner running in the vm instance.

It may be possible to land another PR first that contains only `session_control` related changes, and then land this PR on top of that, but then we will get a broken master if we do this. I think test can still pass but real code will break. It won't be a problem if you managed to merge both PRs before anyone fork the HEAD branch through. 

I will create a smaller PR that only consists of session_control changes to see if test will break.

chaselatta replied:
session_control should not care about the observatory port.

For port discovery, we are migrating to using inspect to discover the port since it is much more reliable than the hub. There will be a tool built into ffx which does this for you so it might make sense to hold off for now.

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -633,18 +676,25 @@ class FuchsiaDevice extends Device {
-    final List<int> ports = <int>[];
+    final Map<int, List<String>> portUrlMap = <int, List<String>>{};
+
+    // In practice it should be safe to assume that there are only one port
```
I don't understand. Is the concern that there might be multiple instances of the same component running?

michaellee8 replied:
Therotically it may be possible for `ls /hub/c/flutter_aot_runner.cmx/35411/out/debug/vmservice-port` to output multiple directories hence multiple ports, although in practice it is very unlikely to happen. Still the code here adddress that by having each port as a key with the same value.

chaselatta replied:
The value returned from `/hub/c/flutter_aot_runner.cmx/35411/out/debug/vmservice-port` will only ever report one port even if there are multiple components running in it. There are situations where you would find multiple runners but that is rare.

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -653,10 +703,36 @@ class FuchsiaDevice extends Device {
         }
       }
+
+      // Get component urls
```
You might find tests easier to write if you pull some of this logic out into separate helper methods. They can be annotated `@visibleForTesting` to make them visible only to tests and other code in this Dart library.

michaellee8 replied:
Agreed, I will refactor this.

zanderso commented:
> I can see some tool_tests failing, but I am not sure if that was my problem. I don't have code I changed are related to those failing tests.

The failing tests that I see do look related to this change. For example there is a failure in:
```
test/general.shard/fuchsia/fuchsia_device_start_test.dart: Fuchsia app start and stop:  fail when cant get host address
```

michaellee8 committed:
fix: typo

Co-authored-by: Zachary Anderson <zanderso@users.noreply.github.com>

michaellee8 commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -266,6 +251,22 @@ class FuchsiaDevice extends Device {
+  ///
+  /// If the device is running a workstation build, `session_control` should be
+  /// used to launch apps, otherwise `tiles_ctl` should be used.
+  Future<bool> isSession() async {
```
Agreed, will refactor this.

michaellee8 replied:
Refactored in 7b975d4530249b1ecc781df4d8020531d7c58eb3.

michaellee8 commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -304,6 +305,15 @@ class FuchsiaDevice extends Device {
+    // Populate isSession first
+    _isSession = await isSession();
+
+    if (_isSession) {
```
Agreed, will refactor this.

michaellee8 replied:
Refactored in 7b975d4530249b1ecc781df4d8020531d7c58eb3.

michaellee8 committed:
refactor: more idiomatic isSession

michaellee8 committed:
docs: add comments about dart ipv6 parsing issues

michaellee8 committed:
fix: broken tests in CI https://cirrus-ci.com/task/4564968260501504?logs=main#L1755

michaellee8 committed:
fix: broken tests in CI https://cirrus-ci.com/task/6639355927199744?logs=main#L1729

michaellee8 committed:
fix: broken tests in CI https://cirrus-ci.com/task/5670273073545216?logs=main#L1634

michaellee8 committed:
Merge remote-tracking branch 'origin/master' into feat/fuchsia-ermine

michaellee8 committed:
refactor: clean out logic in servicePorts

michaellee8 committed:
test: fix tests

michaellee8 committed:
feat: add support for launching fuchsia app using session_control

michaellee8 committed:
Merge remote-tracking branch 'origin/master' into feat/fuchsia-session-detection

michaellee8 committed:
feat: add code for determining whether using session

michaellee8 committed:
Merge branch 'feat/fuchsia-session-detection' into feat/fuchsia-ermine

# Conflicts:
#	packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

michaellee8 committed:
refactor: clean up after merge

michaellee8 committed:
chore: trigger ci rerun

michaellee8 commented:
@zanderso @chaselatta I have refactored as discussed and all tests is passing now. Please review.

michaellee8 committed:
refactor: remove some unwanted session_control related code

michaellee8 committed:
Merge branch 'feat/fuchsia-session-detection' into feat/fuchsia-ermine

michaellee8 committed:
fix: remove non-overridden method in fake classes

michaellee8 committed:
Merge branch 'feat/fuchsia-session-detection' into feat/fuchsia-ermine

michaellee8 commented:
How about this one? Are there any problems blocking this one from being merged other than #85752 ? I will merge all changes in #85752 into this PR.

chaselatta commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -621,40 +656,82 @@ class FuchsiaDevice extends Device {
+
+  @visibleForTesting
+  Future<Iterable<String>> vmServicePaths() async {
+    final Iterable<String> paths = await findByName('/hub', 'vmservice-port');
```
This can be extremely slow and fragile. It is preferable to use the inspect apis to query for the port.

chaselatta replied:
something along the lines of `iquery -f json show root:vm_service_port --manifest flutter_jit_runner.cmx` will give you the port in a json payload.

chaselatta replied:
or if you want to cover both aot and jit you can do `iquery -f json show "flutter_*_runner.cmx:root:vm_service_port"`.

michaellee8 replied:
Let me try to use the inspect api.

michaellee8 replied:
@chaselatta Do the flutter_tools codebase uses any kind of more structured json parsing library like json_serializable? It is pretty hard to do json parsing on iquery result without one.

zanderso replied:
The tool uses the json parser in `dart:convert`. See https://api.dart.dev/dev/2.14.0-305.0.dev/dart-convert/json-constant.html

michaellee8 replied:
@zanderso So doing some parsing by hand is unavoidable anyway?

zanderso replied:
@michaellee8 I don't know what the level of complexity of the json coming from iquery is, but if you only need to extract a few things, then doing that manually is preferable to introducing a new dependency to the tool.

michaellee8 replied:
@zanderso I still find trying to parse those json returned by iquery is really inconvenient, so I wrote a small helper for it to do it in tree, in https://github.com/flutter/flutter/pull/86620. Please help to review it and decide if it can be accepted. I will do the coding of parsing iquery using it if it is accepted.

chaselatta replied:
@michaellee8 We are currently landing a change to put the VM service port publishing on the component instead of the runner. The CL which is adding this is https://fuchsia-review.googlesource.com/c/fuchsia/+/606062. This means that you will be able to launch your component and query the component for the port instead of having to query the runner. This is part of a larger effort to clean up service port discovery on flutter devices. 

Another approach would be to not do this manually but instead use ffx flutter tunnel to set up the tunnels for you.

michaellee8 committed:
feat: wip, broken

jonahwilliams commented:
@michaellee8  , @chaselatta is OOO for several weeks so we might not be able to make a decision on whether to go with this approach (versus switching to more of the SDK tooling as it is made ready).

michaellee8 commented:
@jonahwilliams Nevermind, so I guess we will stall this PR until Chase is back

Hixie commented:
What is the status of this PR?

Hixie commented:
I'm going to close this PR since it seems to have been abandoned.
